presto-main/src/main/java/com/facebook/presto/security/RefreshingSystemAccessControl.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.security;
+
+import com.facebook.presto.spi.CatalogSchemaName;
+import com.facebook.presto.spi.CatalogSchemaTableName;
+import com.facebook.presto.spi.SchemaTableName;
+import com.facebook.presto.spi.security.Identity;
+import com.facebook.presto.spi.security.Privilege;
+import com.facebook.presto.spi.security.SystemAccessControl;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.airlift.log.Logger;
+
+import java.security.Principal;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+public class RefreshingSystemAccessControl
+        implements SystemAccessControl
+{
+    private static final Logger LOG = Logger.get(RefreshingSystemAccessControl.class);
+
+    static final String REFRESH_PERIOD_SEC = "security.refresh-sec";
+    private static final String DEFAULT_REFRESH_PERIOD_SEC = "300"; // 5 mins
+    static final String REFRESH_ENABLED = "security.refresh-enabled";
+    private static final String DEFAULT_REFRESH = "false";
+    private final ScheduledThreadPoolExecutor service;
+
+    private AtomicReference<SystemAccessControl> delegate = new AtomicReference<>();
```

Comment:
Why not supplier?

Reply:
Supplier of a delegate? I guess you could do that, but generally I expect suppliers to be dynamically generating, whereas atomic is very clear as to what it is doing

Reply:
> I expect suppliers to be dynamically generating

No. Supplier responsibility is to supply the object. There is nothing about the way it is created (it could memoized).